Implement the new module system

kgmt0 opened the pull request:
This is still work-in-progress, but you can use it to load NLS files now. The `alias` option that allows you to control how names are prefixed is also working.

kgmt0 committed:
Extend the parser to support the new libraries syntax

This doesn't actually do anything useful yet.

kgmt0 committed:
Extend the parser to support the new libraries syntax

This doesn't actually do anything useful yet.

kgmt0 committed:
Implement NLS file loading with libraries syntax

Only the "alias" option is supported at this time.

kgmt0 committed:
Implement NLS file loading with libraries syntax

Only the "alias" option is supported at this time.

ERSUCC commented:
Do you have documentation or a sample model to demonstrate the new syntax?

ERSUCC commented:
@kgmt0 I played around with it for a bit, and so far I like the way it works and it seems very promising! Here are some issues I noticed:

1. If you import a library and then just use its name without the colon and a command after it, the error message is "Nothing named xyz has been defined." I would expect it to be something along the lines of "xyz is a library, not a command."
2. Imported libraries don't show up under the Included Files dropdown. This is probably up for debate but personally I would like it if they showed up there so I could easily open them.
3. `alias` is used like a keyword as far as I can tell, but the compiler does not identify it as a keyword. This means that it isn't green and bolded like `__includes` or `libraries`, and you can use `alias` as an alias.
4. Including a library in itself causes infinite recursion and the app freezes. This should be caught before it happens and a relevant error should be thrown.
5. Included libraries are not private to the file in which they are included, which seems undesirable to me. For example, if I include library `test` in the Code tab, and I include library `test2` in library `test`, I can now access library `test2` from the Code tab.
6. No error is thrown if you include the same library twice. Similarly, no error is thrown if you include a library once, and then include a different library that is aliased to the same name as the first library. This latter case is problematic, because both libraries are then included, and any command name conflicts become ambiguous. After messing around with it, I discovered that in that case, the version of the command from the first included library is used, and all other commands with the same name are silently ignored.

Let me know if I need to clarify anything I said.

kgmt0 commented:
Thanks for testing!

>     1. If you import a library and then just use its name without the colon and a command after it, the error message is "Nothing named xyz has been defined." I would expect it to be something along the lines of "xyz is a library, not a command."

Hmm. I didn't consider having a special error message for that because libraries are not (currently) a value, but I can see how users would find it helpful. Would love to hear what other people think as well.

>     2. Imported libraries don't show up under the Included Files dropdown. This is probably up for debate but personally I would like it if they showed up there so I could easily open them.

I agree. I plan to implement this but haven't got to it yet.

>     3. `alias` is used like a keyword as far as I can tell, but the compiler does not identify it as a keyword. This means that it isn't green and bolded like `__includes` or `libraries`, and you can use `alias` as an alias.

It's a bit different from other keywords because it only has a special meaning within the context of the libraries syntax. There's no problem if the user wants to name something "alias", or even use the word "alias" as an alias.

>     4. Including a library in itself causes infinite recursion and the app freezes. This should be caught before it happens and a relevant error should be thrown.

Good catch! I'll fix that.

>     5. Included libraries are not private to the file in which they are included, which seems undesirable to me. For example, if I include library `test` in the Code tab, and I include library `test2` in library `test`, I can now access library `test2` from the Code tab.

There's a (not yet implemented) `define-libraries` syntax that you'll be able to use to list exactly the symbols you want to export. Once that's implemented, I plan to make symbols private by default.

>     6. No error is thrown if you include the same library twice. Similarly, no error is thrown if you include a library once, and then include a different library that is aliased to the same name as the first library. This latter case is problematic, because both libraries are then included, and any command name conflicts become ambiguous. After messing around with it, I discovered that in that case, the version of the command from the first included library is used, and all other commands with the same name are silently ignored.

Good point. I'll add some checks for that.

kgmt0 commented:
> Do you have documentation or a sample model to demonstrate the new syntax?

I sent you a link to the design document on Slack.

kgmt0 committed:
Check for duplicate library imports

kgmt0 committed:
Check for duplicate library imports

kgmt0 committed:
Detect loops in library imports

kgmt0 committed:
Detect loops in library imports

kgmt0 committed:
Update syntax for importing libraries

Change the keyword from "libraries" to "library" and make it so that
only one library is accepted at a time.

kgmt0 committed:
Update syntax for importing libraries

Change the keyword from "libraries" to "library" and make it so that
only one library is accepted at a time.

kgmt0 committed:
Properly implement findLibraries()

The previous implementation was a placeholder and didn't actually work.

kgmt0 committed:
Properly implement findLibraries()

The previous implementation was a placeholder and didn't actually work.

kgmt0 committed:
Display files imported with the libraries syntax in include files drop down

kgmt0 committed:
Display files imported with the libraries syntax in include files drop down

kgmt0 committed:
Extend the parser to support the define-library syntax

This doesn't actually do anything useful yet.

kgmt0 committed:
Extend the parser to support the define-library syntax

This doesn't actually do anything useful yet.

kgmt0 committed:
Use the keyword's token when constructing StructureDeclarations.Library

This is more consistent with how other things are parsed. Previously,
the token belonging to the name of the library was used.

kgmt0 committed:
Use the keyword's token when constructing StructureDeclarations.Library

This is more consistent with how other things are parsed. Previously,
the token belonging to the name of the library was used.

kgmt0 committed:
Refactor libraries code in StructureParser

In the section of the code where libraries are handled, there's already
a check to ensure that the list of libraries is not empty. Because of
this, it's not necessary to use headOption to access the head of the
list safely. This replaces headOption to head where applicable and apply
some simplifications.

kgmt0 committed:
Refactor libraries code in StructureParser

In the section of the code where libraries are handled, there's already
a check to ensure that the list of libraries is not empty. Because of
this, it's not necessary to use headOption to access the head of the
list safely. This replaces headOption to head where applicable and apply
some simplifications.

kgmt0 committed:
Remove unreachable match case in StructureConverter

kgmt0 committed:
Update ProceduresMap to use filename as a part of the key

This makes it possible to different procedures with the same name in
different modules.

kgmt0 committed:
Update ProceduresMap to use filename as a part of the key

This makes it possible to different procedures with the same name in
different modules.

kgmt0 committed:
Implement namespacing for procedures in modules

kgmt0 committed:
Implement namespacing for procedures in modules

kgmt0 committed:
Make filename in FrontEndProcedure an Option[String]

kgmt0 committed:
Make filename in FrontEndProcedure an Option[String]

kgmt0 committed:
Rename the "library" syntax to "import"

Also exclude globals, breeds, and breed variables from being imported.

kgmt0 committed:
Rename the "library" syntax to "import"

Also exclude globals, breeds, and breed variables from being imported.

kgmt0 committed:
Add an alias field to Procedure

Add a field that lists different aliases of a procedure to different
variations of Procedure. This allows CompilerResult to generate the
correct procedure map from lists of procedures.

kgmt0 committed:
Add an alias field to Procedure

Add a field that lists different aliases of a procedure to different
variations of Procedure. This allows CompilerResult to generate the
correct procedure map from lists of procedures.

kgmt0 committed:
Allow importing modules from a module

kgmt0 committed:
Allow importing modules from a module

kgmt0 committed:
Add module field to Procedure

Also set the module field of procedures imported with __includes to
None since they're supposed to be in the namespace of the model.

kgmt0 committed:
Add module field to Procedure

Also set the module field of procedures imported with __includes to
None since they're supposed to be in the namespace of the model.

kgmt0 committed:
Rename define-library to export

kgmt0 committed:
Rename define-library to export

kgmt0 committed:
Add checks for name conflicts caused by importing modules

kgmt0 committed:
Add checks for name conflicts caused by importing modules

ERSUCC commented:
Related issue: #1654

kgmt0 committed:
Implement selective export of procedures using the export syntax

kgmt0 committed:
Implement selective export of procedures using the export syntax

kgmt0 committed:
Rename currentLibrary to currentImport in StructureParser

kgmt0 committed:
Rename currentLibrary to currentImport in StructureParser

kgmt0 committed:
Remove version number from the export syntax

kgmt0 committed:
Remove version number from the export syntax

kgmt0 committed:
Rename processedLibraries to processedImports in StructureParser

kgmt0 committed:
Rename processedLibraries to processedImports in StructureParser

ERSUCC commented on netlogo-core/src/main/compile/api/Backifier.scala:
```diff
@@ -7,6 +7,6 @@ import org.nlogo.{ core, nvm }
 trait Backifier {
-  def apply(procedures: ListMap[String, nvm.Procedure], c: core.Command): nvm.Command
-  def apply(procedures: ListMap[String, nvm.Procedure], r: core.Reporter): nvm.Reporter
+  def apply(procedures: ListMap[Tuple2[String, Option[String]], nvm.Procedure], c: core.Command): nvm.Command
```
Is there a reason why you use `Tuple2[T, U]` instead of `(T, U)`?

kgmt0 replied:
Not really. I saw `Tuple2` being used somewhere else so I thought I would do the same thing.

ERSUCC replied:
I defer to @TheBizzle for the final call, but my understanding is that the parenthesis syntax is more idiomatic, and it's what the repo currently uses the most. Not a huge deal, but it may be less familiar to people than the parenthesis syntax.

TheBizzle replied:
Yeah, syntactic sugar preferred.

ERSUCC commented on shared/resources/main/i18n/Errors_en.properties:
```diff
@@ -201,6 +201,11 @@ compiler.MultiAssign.emptyVarList = The list of variables names given to {0} mus
 compiler.StructureParser.includeNotFound = Could not find {0}
+compiler.StructureParser.importNotFound = Could not find {0}
+compiler.StructureParser.importMultipleImports = Attempted to import a module multiple times
+compiler.StructureParser.importLoop = Found a loop in the module import chain
```
Since `loop` is already a thing in NetLogo, it might be better to just reuse the `importMultipleImports` property here. Maybe change the wording slightly, something like "Module {0} has already been imported".

kgmt0 committed:
Change error message for import loops

Change the message to "Module {0} has already been imported".

kgmt0 committed:
Change error message for import loops

Change the message to "Module {0} has already been imported".

kgmt0 committed:
Show transitively imported files in the included files dropdown

kgmt0 committed:
Show transitively imported files in the included files dropdown

kgmt0 committed:
Implement parser for nlproject files

kgmt0 committed:
Implement parser for nlproject files

kgmt0 committed:
Make the import syntax optionally take a package name

Also update the module loading mechanism to use package metadata file to
find module files.

kgmt0 committed:
Make the import syntax optionally take a package name

Also update the module loading mechanism to use package metadata file to
find module files.

kgmt0 committed:
Implement package installation

This makes it possible to install packages through the extension
manager.

kgmt0 committed:
Implement package installation

This makes it possible to install packages through the extension
manager.

kgmt0 committed:
Fix relative imports from inside packages

Previously, relative imports only tries to load module files relative to the
CWD. Fixes this so that it tries to load files relative to the
importing module instead.

kgmt0 committed:
Fix relative imports from inside packages

Previously, relative imports only tries to load module files relative to the
CWD. Fixes this so that it tries to load files relative to the
importing module instead.

kgmt0 committed:
Update library repository URL for testing (DON'T MERGE)

Remove this commit after testing is done.

kgmt0 committed:
Make include files dropdown work with the new import syntax

kgmt0 committed:
Make include files dropdown work with the new import syntax

kgmt0 committed:
Make colon a token

For identifiers that contain colons, its tokens will be merged so that
they can be processed correctly by Namer.

kgmt0 committed:
Make colon a token

For identifiers that contain colons, its tokens will be merged so that
they can be processed correctly by Namer.

kgmt0 committed:
Use backticks instead of underscore to handle handle name conflicts

Instead of prefixing the name with an underscore to avoid conflicts with
reserved keywords, surround it with backticks instead.

kgmt0 committed:
Use backticks instead of underscore to handle handle name conflicts

Instead of prefixing the name with an underscore to avoid conflicts with
reserved keywords, surround it with backticks instead.

kgmt0 committed:
Use stripMargin for multiline strings

Use stripMargin so that multiline strings used in test cases can be
indented properly.

kgmt0 committed:
Use stripMargin for multiline strings

Use stripMargin so that multiline strings used in test cases can be
indented properly.

kgmt0 committed:
Switch to the new import syntax

The new syntax looks something like this "import foo:bar as baz".

kgmt0 committed:
Switch to the new import syntax

The new syntax looks something like this "import foo:bar as baz".

kgmt0 committed:
Display an error when a module contains things other than procedures

kgmt0 committed:
Display an error when a module contains things other than procedures

kgmt0 committed:
Display an error when extensions are used inside a module

kgmt0 committed:
Display an error when extensions are used inside a module

kgmt0 committed:
Remove module name from the export syntax

kgmt0 committed:
Remove module name from the export syntax

kgmt0 committed:
Implement the new import syntax

There are 2 basic forms in the new syntax. The first form allows
importing one or more modules by specifying an import path that more or
less corresponds with the directory structure of the package. If the
import path corresponds to a folder containing multiple modules, all of
them will be made available under appropriate prefixes. The syntax looks
something like this:

import foo:bar:baz

The second syntax allows importing specific identifiers from a module.
Unlike the first syntax, the import path must point to a specific
module. The syntax looks something like this:

import [(baz as qaz) qux] from foo:bar

kgmt0 committed:
Improve error message when non-procedure is found in a module

Make the error messages say what triggered the error.

kgmt0 committed:
Reject uses of export syntax outside modules

Remove the "export" test in StructureParserTests since that is now
expected to fail with this change. There are other tests that involve
the export syntax anyway so this should be fine.

kgmt0 committed:
Fix indentation in StructureDeclarations.scala

kgmt0 committed:
Remove outdated TODOs in LibraryManager.scala and StructureParser.scala

TheBizzle committed:
I18N-en: Reword a couple of new messages

TheBizzle committed:
I18N-de: Provide translations for new module system messages

TheBizzle committed:
Refactor: Tidy up some of the new module system code

* Give curly braces some "breathing room"
* Avoid curly braces when parentheses will suffice
* In Scala, refer to the type as `(A, B)`, instead of `Tuple2[A, B]`
* Favor string interpolation over string concatenation
* Newlines, spacing, indentation

TheBizzle committed:
Refactor: Unify some triplicated code

ERSUCC committed:
Infrastructure: Use extensions API version 7.1
